compiler/src/iree/compiler/Dialect/Util/IR/UtilAttrs.cpp

```diff
@@ -345,6 +356,99 @@ serializeGenericFloatElements(DenseFPElementsAttr attr,
   return success();
 }
 
+template <typename elementType>
+static void serializeGenericIntegerResourcElementsAPInt(std::optional<ArrayRef<elementType>> attrData,
+                                                        unsigned numBits,
+                                                        llvm::support::endianness endian,
+                                                        llvm::raw_ostream &os) {
+  for (int i = 0; i < attrData->size(); i++) {
+    auto val = (*attrData)[i];
+    APInt newVal(numBits, val);
+    elementType rawValue = llvm::support::endian::byte_swap<elementType>(
+        newVal.extractBitsAsZExtValue(numBits, 0), endian);
+    os.write((const char *)&rawValue, sizeof(rawValue));
+  }
+}
+
+template <typename elementType>
+static void serializeGenericFloatResourcElementsAPFloat(std::optional<ArrayRef<elementType>> attrData,
+                                                        unsigned numBits,
+                                                        llvm::support::endianness endian,
+                                                        llvm::raw_ostream &os) {
+  for (int i = 0; i < attrData->size(); i++) {
+    auto val = (*attrData)[i];
+    APFloat newVal(numBits, val);
+    elementType rawValue = llvm::support::endian::byte_swap<elementType>(
+        newVal.bitcastToAPInt().extractBitsAsZExtValue(numBits, 0), endian);
+    os.write((const char *)&rawValue, sizeof(rawValue));
+  }
+}
+
+template <typename elementType, unsigned numBits = sizeof(elementType) * 8>
+static LogicalResult
+serializeGenericIntegerResourceElements(Location loc,
+                                DenseResourceElementsAttr attr,
+                                llvm::support::endianness endian,
+                                llvm::raw_ostream &os) {
+  unsigned bitWidth = attr.getType().getElementTypeBitWidth();
+  switch (bitWidth) {
+    case 16:
```

Comment:
i8 is supported, but regarding the subbyte implementation that you requested, I actually wrote a whole serialization code for that case. But, then after some digging into DenseResource realized it's not possible. DenseResource only supports DenseBoolResourceElementsAttr, DenseI8ResourceElementsAttr, DenseI16ResourceElementsAttr, DenseI32ResourceElementsAttr, DenseI64ResourceElementsAttr, DenseUI8ResourceElementsAttr, DenseUI16ResourceElementsAttr, DenseUI32ResourceElementsAttr, DenseUI64ResourceElementsAttr, DenseF32ResourceElementsAttr, and DenseF64ResourceElementsAttr. This is different than DenseElements where we can just have a DenseIntElementsAttr which is why subbyte serialization can be supported with the PackedWriter for DenseElements.

Reply:
To add on, my plan in torch-mlir is to check the ElementTypeBitWidth for the Dense Attribute, and if it less than 8 we proceed with DenseElementsAttr. Otherwise, we go with DenseResourceElementsAttr.